Code and Implementation for #15

takposha opened the pull request:
Fixes #15 

`questionGenerator.py` handles the question generation aspect of the pipeline. It uses a `QuestionData` class to store the relevant information. It is similarly structured to `topicExtractor.py` to be easier to follow.

Changes to the ReadMe, `.env.sample`, and `config.py` files support this final part.

Simple functions to print the topics and questions have also been implemented. 

This should cover all of the code I demonstrated on Friday last week.

takposha committed:
Adds code for #15

`questionGenerator.py` handles the question generation aspect of the pipeline. It uses a `QuestionData` class to store the relevant information. It is similarly structured to `topicExtractor.py` to be easier to follow.

Changes to the ReadMe, `.env.sample`, and `config.py` files support this final part.

Simple functions to print the topics and questions have also been implemented.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,272 @@
+            printAndLog(f"Topic: {topic}")
+            printAndLog(f"Insert Point: {self.dominantTopics[topic]['End'].strftime('%H:%M:%S')}")
+            question = f"Question: {parsedResponse['question']}"
+            answers = f"Answers: {'\n\t '.join(parsedResponse['answers'])}"
```
This line throws error saying `"SyntaxError: f-string expression part cannot include a backslash" `
after modifying something like this 
`answers = "Answers: " + '\n\t'.join(parsedResponse['answers'])` was OK. 
did you encounter this?

takposha replied:
No, this is not an error for me. Are you on an older version of Python, maybe? I think this was only implemented post-Python 3.10.

pushyamig replied:
This is my version `Python 3.10.9`

takposha replied:
My apologies. I looked it up and it appears to be newer than I thought, it was implemented in Python 3.12. Should I change the syntax to keep it compatible with previous versions or enforce a newer Python version in the requirements?

pushyamig replied:
Yes, please. Otherwise we have to start the process running in Docker, which is aimed for future time.

takposha replied:
To clarify, should I change the syntax, or update the `requirements.txt`?

pushyamig replied:
>update the requirements.txt?

why? this is related to Python version and not related to packages.

To keep it simple change syntax

takposha replied:
I see. Thank you for the clarification.

takposha committed:
Syntax updated in `questionsGenerator.py` to remove backlsash usage in f-string, which is not supported in Python <3.12

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,272 @@
+        self.config = config
+
+        self.clusteredTopics = getClusteredTopics(videoData, topicModeller)
+        self.dominantTopics = getDominantTopic(self.clusteredTopics)
```
Log top 10 or 5 topic returned in both cases. The object that is returned looks complicated so logging relevant info is useful

takposha replied:
This has been addressed in [the new commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/16/commits/dc537b7992d22ba4830300124c4c57ff326eedeb).

pushyamig replied:
Thanks! Logging certain info while the process is running is useful as intermediate step

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,272 @@
+    def __init__(self, config, videoData, topicModeller, load=True):
+        self.config = config
+
+        self.clusteredTopics = getClusteredTopics(videoData, topicModeller)
```
What will happen if `self.clusteredTopics` returns empty list? should be allow it to continue ?

takposha replied:
This has been addressed in [the new commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/16/commits/dc537b7992d22ba4830300124c4c57ff326eedeb).

In theory, I do not think it should be possible for this to ever return empty, as BERTopic should have failed beforehand, which would have been caught in checks there. But as you mentioned, I should account for the 'unhappy' paths as well!

pushyamig replied:
Thanks! Logging certain info while the process is running is useful as intermediate step

takposha committed:
Added logging messages for `clusteredTopics` and `dominantTopics`, as well as checks for empty DataFrames being returned.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,291 @@
+        self.clusteredTopics = getClusteredTopics(videoData, topicModeller)
+        self.dominantTopics = getDominantTopic(self.clusteredTopics)
+        self.relevantText, self.questionText = getTextAndQuestion(
+            self.dominantTopics, videoData
```
Log the question and text data from the output.

takposha replied:
Addressed in the [new Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/16/commits/be278feff6e0c397bf70603df9d69149c0fcf026).

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,291 @@
+        if load:
+            self.loadQuestionData()
+        else:
+            self.OpenAIChatBot = OpenAIBot(self.config)
```
Is there any reason why you are using OpenAIBot Instead of LangChainBot? As I understood you are using LangChain for topic extraction.

takposha replied:
This is because this is a much simpler task than the Topic Extraction part. Using LangChain in BERTopic lets BERTopic handle a lot of the complexities in how the topics are generated by leveraging LangChain, as it passes keywords, relevant documents, and such to get the human-interpretable topics.

In the case of Question generation, we only need to send one query per question we need to generate, and we get much finer control on the prompt directly because of it (LangChain is just a complicated way of building the prompt and storing information at its fundamental level as well, it just hides it from us through abstraction). So in this case, it made sense to stick to an OpenAIBot for our call rather than using LangChain.

pushyamig replied:
Please document that here and add link for further reading if there is any

takposha replied:
Addressed in the [new Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/16/commits/be278feff6e0c397bf70603df9d69149c0fcf026).

There isn't a link to document my exact implementation reasoning, so I put my explanation down there instead, but I did add a link to back up my reasoning for how LangChain is just a system atop OpenAI's prompting calls.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,291 @@
+        else:
+            self.OpenAIChatBot = OpenAIBot(self.config)
+            self.getQuestionData()
+            config.questionTokenCount = self.OpenAIChatBot.tokenUsage
```
are you trying to override the config Value? that's my reading from here. We usually don't override value related to configuration/env variable but just get then when running the process.

takposha replied:
For the token count? Yes, it's kind of being overridden here. 
The Config object is a persistent object that can store information throughout the pipeline, so that is why I was passing the token value to it as a way to store it. 
This does fall under Issue #17, where I would like to improve the way token usage might be tracked.

pushyamig replied:
@takposha , The problem I am trying to raise here is `Config` Class main purpose seems to be holding the `.env` variable. So you can look at this class as holder of a static info that only changes once before the process/script runs.

The movement you feed other variables that are irrelevant into this Object then it's not a Config Class anymore.

What you can do here is

1. Config Class only stores `.env` file values
2. Move all the static info variables to beginning `config.py` of the  of the  and grab those variable elsewhere by importing. Below are the variable that fits this. 
    1. `self.openAIParams`, `self.captionsFolder`, `self.saveFolder`
    ```
     `config.py`
     CAPTIONS = "Captions"
      # when Importing do this say to `topicExtractor.py`
     from config import CAPTIONS
    # this solution works for single variable but I want you think through on the dict object, `openAPIParams` in similar context
    ```
3.  for variables `topicTokenCount`, `questionTokenCount` in the Config Class these as per my reading the current code it is only used that `QuestionData` and `TopicModeller`and only for logging purpose. So these need to be class variable for respective class and not part of the `Config` Class. If you see a purpose in future code other than logging, I feel you need pass the Class instance to grab the values.

pushyamig replied:
The code for Config.py was introduce in last PR https://github.com/tl-its-umich-edu/annoto-gai/pull/14. The PR was huge, so It did not caught my attention. This is one of the reason's why PR should be small or sizable for easy review.

takposha replied:
Hmm, I see. So what if I do the following:
I have three main classes: 
1. configVars for setting up environmental variables,
2. staticVars for storing all static variables that need to be passed,
3. trackingVars for storing any variables that are used to track and update stuff.

I put all of it together in one wrapper class to pass to all of the other functions (What do I call this, and I should maybe rename `config.py` to something else). Will this be a way to handle the different variables we are using in an effective manner?

takposha replied:
I will also create functions with the `TopicModeller` and `QuestionData` classes that return the token usage which can then be assigned to the variables that are tracking and logging these values. That will help with #17, so I will implement that in a smaller PR for that issue.

pushyamig replied:
`configVars` : This could be called as official Config class
`staticVars`: I don't know if you want to create class for static variable. These are simply be put in `contancts.py`
`trackingVars`: Separate it a different Object, don't know if you want a class for holding 2 variable. the topicTokenCount is part of `TopicModeller` and is available to `QuestionData`. So you could just create a simple dict object and pass it along. I will leave up to you.

pushyamig replied:
`callMaxLimit` should be part of `TopicModeller` and not part of `Config` class. just mentioning it here as well. I don't see this being used elsewhere

takposha replied:
Do we need to have the call limit ever set as an environmental value in the future, actually? I chose 5 as a default value, but if it doesn't need to be changed, I will move it into the actual classes where it is used only.

pushyamig replied:
Ideally yes, But I let it pass now. But I did notice it lasttime. I feel 3 should be good number than 5

takposha replied:
Noted. I will set it up as a fixed value in the class for now, and make an environmental value later if needed.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,291 @@
+        else:
+            self.OpenAIChatBot = OpenAIBot(self.config)
+            self.getQuestionData()
+            config.questionTokenCount = self.OpenAIChatBot.tokenUsage
```
If you are keeping track of token usage, this needs to Logged as info.

takposha replied:
Addressed in the [new Commit](https://github.com/tl-its-umich-edu/annoto-gai/pull/16/commits/be278feff6e0c397bf70603df9d69149c0fcf026).

takposha committed:
Added logging information for generated Query and Text data in `questionGenerator.py`

Added logging information for token usage in `topicExtractor.py` and `questionGenerator.py`

Added documentation to explain OpenAI call use over LangChain in `questionGenerator.py`

Added a check to automatically overwrite Generated Question data if Topic Model data is set to be overwritten.

Added a try-except catch for failing to save the Topic Model in `topicExtractor.py`.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,318 @@
+        load (bool, optional): Whether to load existing question data. Defaults to True.
+    """
+
+    def __init__(self, config, videoData, topicModeller, load=True):
```
I would avoid writing the huge business logic when instantiate the class.  Like loading and saving data, clustering , filtering dominats topic and instantiating OpenAIBot. That's a lot.

Create a function like some kind of orchestrator  after instantiate the class that handle all the workflow going with Question genration

pushyamig commented on topicExtractor.py:
```diff
@@ -10,11 +10,13 @@
 
```
My Thought shared here https://github.com/tl-its-umich-edu/annoto-gai/pull/16/files#r1611987954 can be applied to this class well.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,318 @@
+            response = response.strip("` \n")
+            if response.startswith("json"):
+                response = response[4:]
+            parsedResponse = json.loads(response)
```
This step here, underneath `loads` implementation raises error like for example `JSONDecodeError`. Wheneven you are using a pkg object functions need to account for how are errors are handled. Unhappy Path consideration.

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,318 @@
+        filteredTopics.groupby(
+            (filteredTopics["Topic"] != filteredTopics["Topic"].shift()).cumsum()
+        )
+        .agg(
```
This function implementation is raising error. So Understand if that error is wrap in `try-catch`

takposha replied:
Do you have any more info on what video caused this error? I will wrap the function calls in a `try-catch` but I'd also like to replicate the error itself, it might indicate something needs to be adjusted for topic extraction.

pushyamig replied:
New Quizzes Video

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,318 @@
+        """
+        Loads the question data from a file.
+        """
+        self.responseData = dataLoader(
```
Surround by `try-catch` here or at `utils` in `pickle.load` with raise an error

pushyamig commented on questionGenerator.py:
```diff
@@ -0,0 +1,318 @@
+        """
+        Saves the question data to a file.
+        """
+        dataSaver(
```
Surround by try-catch here or at utils in `pickle.dump` with raise an error

pushyamig commented on captionsProcessor.py:
```diff
@@ -1,22 +1,24 @@
+    questionData = retrieveQuestions(config, videoData, topicModeller)
+    questionData.printQuestions()
 
 if __name__ == "__main__":
```
I see warning 
```
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
```
Probably I might over looked with before PR.  Can this be fixed?

takposha replied:
Is this only a warning? It seems to be something happening as the data is tokenized to be used for LangChain, but I haven't recorded this happening on my system, so I am not sure what is triggering it.

takposha committed:
Fixes for mutliple comments in PR#16

`config.py` has been renamed to `configData.py` with the class Config renamed and adjusted to store only .env data, and static variables have been seperated.

Multiple `try-catch` blocks have been applied across all scripts to better catch possible errors

The data classes for Transcript and Question data have been reworked to remove business logic from the `__init__` function, and been reworked to have similar behavior.

takposha committed:
Refactor config import and corrected initialization in captionsProcessor.py

takposha commented:
The recent commit should address all issues raised in comments so far in this PR, however, two comments could not be completely resolved:

> This function implementation is raising error. So Understand if that error is wrap in `try-catch`

A `try-catch` block has been implemented, but the actual error could not be replicated with the New Quizzes Video. Further testing might be needed to isolate this error. The Jupyter Notebook in the commit outputs the results from using the video to get Question Data.

> I see warning
> 
> ```
> ...
> 	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> ```
> 
> Probably I might over looked with before PR. Can this be fixed?

This warning also could not be replicated. It appears to be occurring during the fitting of topics in `topicExtractor.py`, possibly as the data is tokenized for use by BERTopic. Exact cause of warning is unknown.

takposha committed:
Added 'warn' as log level option to denote `WARNING` in `printAndLog` in `utils.py`

takposha committed:
bugfix: Fixed folder creation when not present in `configData.py` and added a `try-catch` for it.

takposha committed:
Reduced logging verbosity for the Query sent to the OpenAI API in `questionGenerator.py`
